### ISSUE TITLE
Self Hosted Ollama Integration

### ISSUE THREAD

**Issue by awptechnologies** [activepieces/activepieces#3307]
It would be amazing if there was a way to incorporate self hosted llama. Giving users the ability to use ollama gives us the ability to really sculpt activepieces AI responses to our liking.

**Comment by abuaboud** [comment]
I am closing this issue, feel free to reopen If the question is not answered

**Comment by icodeforyou-dot-net** [comment]
@abuaboud 

LocalAI does NOT support Ollama as a backend sadly. Please reopen this. 

Ollama is also a much bigger thing now. So it would be really good to have Ollama supported directly. Basic integration shouldn't be hard to do I feel.

There is already something on your community platforms that went in the right direction I feel:

https://community.activepieces.com/t/ollama-or-groq-or-all-the-other-llms-that-use-the-same-api-format-as-openai/4215

**Comment by MarekSurma** [comment]
I would like to mention that it is possible to access ollama server with LocalAI piece. Please use the Custom API call like on the attached images. This is only for the /generate endpoint but it shows that working with ollama through LocalAI is possible.
![image](https://github.com/activepieces/activepieces/assets/3234997/aeaac5e0-b6c8-4fa3-a6a1-3e0b7cd3d85e)
![image](https://github.com/activepieces/activepieces/assets/3234997/6e7b6244-6f0f-46d3-8166-a47dee716627)
![image](https://github.com/activepieces/activepieces/assets/3234997/f6469cf9-b07e-4b0f-8c45-8a2cc476457d)
![image](https://github.com/activepieces/activepieces/assets/3234997/ad5e44a8-cb0e-4676-95ae-f5695d3f34d4)

**Comment by icodeforyou-dot-net** [comment]
> I would like to mention that it is possible to access ollama server with LocalAI piece.

Yes, that is a workaround I wasn't aware of, if only a bit of a tedious one since you now need to run two services side by side. Having a direct ollama integration via a separate piece would still be great.

**Comment by remon-rakibul** [comment]
> I would like to mention that it is possible to access ollama server with LocalAI piece. Please use the Custom API call like on the attached images. This is only for the /generate endpoint but it shows that working with ollama through LocalAI is possible.
> 
> ![image](https://private-user-images.githubusercontent.com/3234997/338897055-309806b8-a984-4994-a3a2-8aabfcb7124b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwMjk1NTAsIm5iZiI6MTcyNjAyOTI1MCwicGF0aCI6Ii8zMjM0OTk3LzMzODg5NzA1NS0zMDk4MDZiOC1hOTg0LTQ5OTQtYTNhMi04YWFiZmNiNzEyNGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MTFUMDQzNDEwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTExMGIxNjQwN2ZmNmJiOWEzMTcxNjhhZmVjZjdmMGRkMWM0NzA0MmU0MWVmM2YwMmEzMzEzNzY2OWFmOTEyOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.XYuF_VhGHX-gIoBirh6rg669GWvx2Yw5VDURAAycLJo) ![image](https://private-user-images.githubusercontent.com/3234997/338897405-09c530b6-65bf-4907-ab9b-366f296c0500.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwMjk1NTAsIm5iZiI6MTcyNjAyOTI1MCwicGF0aCI6Ii8zMjM0OTk3LzMzODg5NzQwNS0wOWM1MzBiNi02NWJmLTQ5MDctYWI5Yi0zNjZmMjk2YzA1MDAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MTFUMDQzNDEwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDc3MDc0ZTc0ZmVhODczN2Y3NzAzNTUzOWQ3N2IxOWZlOWU1NjVlNGE0N2YwYzM4MWI4OGEyZTdjMjhhMGNjMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.4vUxVd3x6ZZHQ-oLgA9TJt07MfeEJ0nLEFYC8_GBEbg) ![image](https://private-user-images.githubusercontent.com/3234997/338897608-27847694-f37c-48ac-88ed-dcf6856e04ba.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwMjk1NTAsIm5iZiI6MTcyNjAyOTI1MCwicGF0aCI6Ii8zMjM0OTk3LzMzODg5NzYwOC0yNzg0NzY5NC1mMzdjLTQ4YWMtODhlZC1kY2Y2ODU2ZTA0YmEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MTFUMDQzNDEwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NzJiODZkNzVmZGQzZDIwYzZiYTRhMzUzZTZiNTBmNWIwYjg1YmEyNTE2ZWVkZDlmYmY4OWExNzE3Mzk5OTE2MyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Ywz5aIkh8rs732OrLqRU4KOF0PlttfWVenR2OiUj7K0) ![image](https://private-user-images.githubusercontent.com/3234997/338897808-aeecc7a9-24d1-4816-8c26-bc3d4139305f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwMjk1NTAsIm5iZiI6MTcyNjAyOTI1MCwicGF0aCI6Ii8zMjM0OTk3LzMzODg5NzgwOC1hZWVjYzdhOS0yNGQxLTQ4MTYtOGMyNi1iYzNkNDEzOTMwNWYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MTFUMDQzNDEwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTY1ZGE5MzFkNjUxNzMzNWU2OGNmY2NmMTc0NTdkOTExYzlmOTFhYTZmYzI1NGVhOWQxYjk5NjliMDBmOTcyZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.qcFbO9ut3tW_WeJjqt99wa2qI1KYJSHdnNaZd_lx7Lc) ![image](https://private-user-images.githubusercontent.com/3234997/338898073-a81e9e17-7d38-4556-90f4-9bf9f22e868a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwMjk1NTAsIm5iZiI6MTcyNjAyOTI1MCwicGF0aCI6Ii8zMjM0OTk3LzMzODg5ODA3My1hODFlOWUxNy03ZDM4LTQ1NTYtOTBmNC05YmY5ZjIyZTg2OGEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MTFUMDQzNDEwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzFjODc1MDk0MjgyYjdjOWUzMjYyMGRiMmNiYWI3MTk4MTQxODM5Y2MyOTZlZmRiYTkyNzBlYzJjNjZhZjdmYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.8QE2Y4nZuyV0aCF55zK2GsI4ng-TLa62QFXQJoLbWqM)
> 
> And here is the result
> 
> ![image](https://private-user-images.githubusercontent.com/3234997/338887247-ad5e44a8-cb0e-4676-95ae-f5695d3f34d4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwMjk1NTAsIm5iZiI6MTcyNjAyOTI1MCwicGF0aCI6Ii8zMjM0OTk3LzMzODg4NzI0Ny1hZDVlNDRhOC1jYjBlLTQ2NzYtOTVhZS1mNTY5NWQzZjM0ZDQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MTFUMDQzNDEwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Njc0NTkzZmY4NWNmYmI5ZjNkNzllZTZlMzU5NTdhMDU0ODk3N2U2YjU1YTA1Mjc3NmJkNWE3MWJlNWY3OTk3MCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.9mVHp0p8IaP-cqvy78zRxgbSnl94HbfZCKM6fyuj_L8)

I tried the same way you did but I am getting unexpected results.
![Screenshot 2024-09-11 at 12 05 05 PM](https://github.com/user-attachments/assets/c736e33e-e9ce-4c74-9d01-6b5ac77ec1ed)

**Comment by ic4-y** [comment]
Frankly the only real solution is to get a separte Ollama piece. No idea if I ever have the time to make a contribution.

In your case however @remon-rakibul this most likely is a server error. You should check the logs to see what is going on.

**Comment by magnus919** [comment]
Yeah this workaround is academically interesting but not terribly useful for a "no code" platform. There are so many OpenAI-compatible API endpoints out there. Maybe the real fix is to add an option to the OpenAI integration to let the user set a non-default URL?

**Comment by orkutmuratyilmaz** [comment]
Any updates?

**Comment by cchance27** [comment]
I always wonder why projects just don’t default to the OpenAI endpoint as even third parties like Gemini have standardized on it and most local servers and apps expose it

**Comment by hotlong** [comment]
+1

**Comment by zjawj** [comment]
程序员必备！🚀https://code.mayoubang.cn 免费 AI 代码工具，助你效率翻倍，告别加班！

**Comment by kaovilai** [comment]
```
http://host.docker.internal:11434
```
works for me using openai provider with baseurl changed to that which is pointed to ollama on the macos host.

**Comment by Mihai-CMM** [comment]
> ```
> http://host.docker.internal:11434
> ```
> 
> works for me using openai provider with baseurl changed to that which is pointed to ollama on the macos host.

How do you do this?

**Comment by ErroneousBosch** [comment]
The lack of Ollama (or even LiteLLM) Integration is a real detriment, especially since it is already in competitors like Flowise and N8n. I am not sure how @kaovilai was changing the baseurl for OpenAI (maybe the OPENAI_BASE_URL env variable?) but that didn't seem to work for me.

Like lacking Minio support, shows a lack of concern for self-hosting.

**Comment by flex-yeongeun** [comment]
+1

**Comment by kaovilai** [comment]
its in the UI somewhere. there's like two or three places to put openai_base_url or its equivalents.. alas I just moved to n8n for now since its more mature in general and fits my self hosted needs good enough within its license restrictions.

**Comment by kaovilai** [comment]
i know there's at least one spot for the pieces bits, and another place for "system ai" where its not workflow specific.

**Comment by VONVONONE** [comment]
custom api call seem to work for me but need workaround
Lm-stuido work well
ollama need Parse response output 
below are my configuration 

lm-studio
<img width="415" height="520" alt="Image" src="https://github.com/user-attachments/assets/dbc2040c-b4f8-4a86-b935-f7f6b951d4f6" />

ollama (need Parse response output)

<img width="419" height="557" alt="Image" src="https://github.com/user-attachments/assets/518ffbfc-b760-417b-a1f3-fd16bdcf7006" />

**Comment by aadarsh-nagrath** [comment]
I think Ollama would be a great addition as a dedicated piece. The current workarounds are messy and not user-friendly, so a proper integration would make Activepieces much more appealing for self-hosted AI workflows and bring it closer to competitors like n8n. It’s definitely worth supporting.

**Comment by pironev** [comment]
Hi, for ActivePieces installed on local Docker I used **LocalAI (Ask LocalAI)** and for connection to Ollama: http://host.docker.internal:11434/v1, with any api key, (ollama for example)

<img width="299" height="407" alt="Image" src="https://github.com/user-attachments/assets/86db1f2b-efc0-4abc-9e8e-b78769380dc7" />

I'm new to this, but I think it should work fine with any OpenAI style API. You can try it.

**Comment by kiddhustle** [comment]
> > ```
> > http://host.docker.internal:11434
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > works for me using openai provider with baseurl changed to that which is pointed to ollama on the macos host.
> 
> How do you do this?

This is because if the ollama is running within docker, localhost  is a host on the internal docker virtual network.

By default docker dameon will expose the following hostname that maps (via network bridge) to the private internal docker network

 http://host.docker.internal:11434

You can find some information on this here:

https://docs.docker.com/desktop/features/networking/networking-how-tos/#core-networking-how-tos
